modules/objdetect/src/qrcode.cpp

```diff
@@ -57,88 +56,102 @@ void QRDecode::init(Mat src, double eps_vertical_, double eps_horizontal_)
+        {
+            if (bin_barcode_row[x] == future_pixel)
             {
-                next_pixel = bin_barcode.at<uint8_t>(i + 1, y);
-                temp_length++;
-                if (next_pixel == future_pixel)
-                {
-                    future_pixel = 255 - future_pixel;
-                    test_lines.push_back(temp_length);
-                    temp_length = 0;
-                    if (test_lines.size() == 5) { break; }
-                }
+                future_pixel = 255 - future_pixel;
+                pixels_position[index] = x;
+                index++;
             }
+        }
+        pixels_position[index] = width_bin_barcode - 1;
+        index++;
+        for (size_t i = 2; i < index - 4; i+=2)
+        {
+            test_lines[0] = static_cast<double>(pixels_position[i - 1] - pixels_position[i - 2]);
+            test_lines[1] = static_cast<double>(pixels_position[i    ] - pixels_position[i - 1]);
+            test_lines[2] = static_cast<double>(pixels_position[i + 1] - pixels_position[i    ]);
+            test_lines[3] = static_cast<double>(pixels_position[i + 2] - pixels_position[i + 1]);
+            test_lines[4] = static_cast<double>(pixels_position[i + 3] - pixels_position[i + 2]);
 
-            // --------------- Compute vertical lines --------------- //
-
-            if (test_lines.size() == 5)
-            {
-                double length = 0.0, weight = 0.0;
+            double length = 0.0, weight = 0.0;
 
-                for (size_t i = 0; i < test_lines.size(); i++) { length += test_lines[i]; }
+            for (size_t j = 0; j < test_lines_size; j++) { length += test_lines[j]; }
 
-                CV_Assert(length > 0);
-                for (size_t i = 0; i < test_lines.size(); i++)
-                {
-                    if (i == 2) { weight += fabs((test_lines[i] / length) - 3.0/7.0); }
-                    else        { weight += fabs((test_lines[i] / length) - 1.0/7.0); }
-                }
+            if (length == 0) { continue; }
+            for (size_t j = 0; j < test_lines_size; j++)
+            {
+                if (j == 2) { weight += fabs((test_lines[j] / length) - 3.0/7.0); }
+                else        { weight += fabs((test_lines[j] / length) - 1.0/7.0); }
+            }
 
-                if (weight < eps_vertical)
-                {
-                    Vec3d line;
-                    line[0] = x; line[1] = y, line[2] = length;
-                    result.push_back(line);
-                }
+            if (weight < eps_vertical)
+            {
+                Vec3d line;
+                line[0] = static_cast<double>(pixels_position[i - 2]);
+                line[1] = y;
+                line[2] = length;
+                result.push_back(line);
             }
         }
     }
     return result;
 }
 
-vector<Point2f> QRDecode::separateHorizontalLines(vector<Vec3d> list_lines)
+vector<Point2f> QRDetect::separateVerticalLines(vector<Vec3d> list_lines)
```

Comment:
replace the signature to `QRDetect::separateVerticalLines(const vector&lt;Vec3d&gt;& list_lines)`. Do not pass `vector&lt;&gt;` by value